File: 199-binary-tree-right-side-view/binary-tree-right-side-view.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function rightSideView(root: TreeNode | null): number[] {
    if (root === null) return []
    
    const result: number[] = []
    const queue: TreeNode[] = [root]

    while (queue.length > 0) {
        const levelSize: number = queue.length

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift()
            if (i === levelSize - 1) result.push(node.val)
            if (node.left !== null) queue.push(node.left)
            if (node.right !== null) queue.push(node.right)
        }
    }

    return result
};

File: 199-binary-tree-right-side-view/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        res = []
        queue = deque([root])

        while queue:
            level_size = len(queue)
            for i in range(level_size):
                current = queue.popleft()
                if i == level_size - 1:
                    res.append(current.val)
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)

        return res
